src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpMetricSerializer.cs

```diff
@@ -0,0 +1,354 @@
+    {
+        writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, fieldNumber, ProtobufWireType.LEN);
+        int dataPointLengthPosition = writePosition;
+        writePosition += ReserveSizeForLength;
+
+        // Casting to ulong is ok here as the bit representation for long versus ulong will be the same
+        // The difference would in the way the bit representation is interpreted on decoding side (signed versus unsigned)
+        writePosition = ProtobufSerializer.WriteFixed64WithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.NumberDataPoint_Value_As_Int, (ulong)value);
+
+        var startTime = (ulong)metricPoint.StartTime.ToUnixTimeNanoseconds();
+        writePosition = ProtobufSerializer.WriteFixed64WithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.NumberDataPoint_Start_Time_Unix_Nano, startTime);
+
+        var endTime = (ulong)metricPoint.EndTime.ToUnixTimeNanoseconds();
+        writePosition = ProtobufSerializer.WriteFixed64WithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.NumberDataPoint_Time_Unix_Nano, endTime);
+
+        foreach (var tag in metricPoint.Tags)
+        {
+            writePosition = WriteTag(buffer, writePosition, tag, ProtobufOtlpMetricFieldNumberConstants.NumberDataPoint_Attributes);
+        }
+
+        if (metricPoint.TryGetExemplars(out var exemplars))
+        {
+            foreach (ref readonly var exemplar in exemplars)
+            {
+                writePosition = WriteExemplar(buffer, writePosition, in exemplar, exemplar.LongValue, ProtobufOtlpMetricFieldNumberConstants.NumberDataPoint_Exemplars);
+            }
+        }
+
+        ProtobufSerializer.WriteReservedLength(buffer, dataPointLengthPosition, writePosition - (dataPointLengthPosition + ReserveSizeForLength));
+        return writePosition;
+    }
+
+    private static int WriteNumberDataPoint(byte[] buffer, int writePosition, int fieldNumber, in MetricPoint metricPoint, double value)
+    {
+        writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, fieldNumber, ProtobufWireType.LEN);
+        int dataPointLengthPosition = writePosition;
+        writePosition += ReserveSizeForLength;
+
+        // Using a func here to avoid boxing/unboxing.
+        writePosition = ProtobufSerializer.WriteDoubleWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.NumberDataPoint_Value_As_Double, value);
+
+        var startTime = (ulong)metricPoint.StartTime.ToUnixTimeNanoseconds();
+        writePosition = ProtobufSerializer.WriteFixed64WithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.NumberDataPoint_Start_Time_Unix_Nano, startTime);
+
+        var endTime = (ulong)metricPoint.EndTime.ToUnixTimeNanoseconds();
+        writePosition = ProtobufSerializer.WriteFixed64WithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.NumberDataPoint_Time_Unix_Nano, endTime);
+
+        foreach (var tag in metricPoint.Tags)
```

Comment:
@CodeBlanch This is a ReadOnlyTagCollection, should we change to for instead of foreach?
Current implementation in OTLP uses foreach.

Reply:
`foreach` is fine for `ReadOnlyTagCollection`! It has this shape...

```csharp
public readonly struct ReadOnlyTagCollection
{
   public Enumerator GetEnumerator() {}
   public struct Enumerator {}   
}
```

Because everything is a `struct` no allocations 😁

This type of thing is also fine...

```csharp
List<int> list = [];
foreach (var item in list) // Uses List<T>.Enumerator which is a struct
{
}
```

The allocations happen when you hit `IEnumerable<T>.GetEnumerator()` which has to return `IEnumerator<T>`.